src/analyses/threadJoins.ml

```diff
@@ -16,18 +16,21 @@ struct
   module G = MustTIDs
   module V = TID
 
-  (* transfer functions *)
-  let return ctx (exp:exp option) (f:fundec) : D.t =
+  let threadexit ctx =
     (
       match ctx.ask CurrentThreadId with
-      | `Lifted tid -> ctx.sideg tid ctx.local
+      | `Lifted tid when ThreadReturn.is_current (Analyses.ask_of_ctx ctx) -> ctx.sideg tid ctx.local
       | _ -> () (* correct? *)
     );
     ctx.local
 
+  (* transfer functions *)
+  let return ctx (exp:exp option) (f:fundec) : D.t = threadexit ctx
+
   let special ctx (lval: lval option) (f:varinfo) (arglist:exp list) : D.t =
     let desc = LibraryFunctions.find f in
     match desc.special arglist with
+    | ThreadExit _ ->  threadexit ctx
```

Comment:
I think it isn't right now, because `threadexit` only side effects for thread returns, but `pthread_exit` may be called from some inner function in the thread and that should still account for a thread return, which currently would be missed.
The correct structure should be the one used by apron analysis, that `ThreadReturn` is only considered in `return` and `ThreadExit` does it unconditionally.

Reply:
Nice catch!

Reply:
I added an example that was unsound before this fix.